text/0000-coroutines.md

```diff
@@ -0,0 +1,485 @@
+- During subsequent invocations, execution resumes immediately after the last executed yield point. 
+  The value of the **yield** expression will be a tuple of the parameters provided by the caller.
+- When execution reaches a **return** statement, or falls off the end of the coroutine body, 
+  the coroutine returns for the last time, the return value being wrapped in `CoResult::Return`.
+- Further attempts to invoke that coroutine shall result in panic.
+
+### Typing
+The types of coroutine signature are subject to normal type inference with a few additional constraints:
+- All return'ed values must be of the same type `R` (same as for regular functions/closures).
+- All yield'ed values must be of the same type `Y`.
+- The tuple of coroutine arguments and the return types of all **yield** expressions must be of the same type `A`. 
+  In other words, `typeof (a1,a2,a3)` == `typeof (b1,b2,b3)` == `typeof (c1,c2,c3)`.
+
+### Hoisting of locals
+Local variables in coroutine body, whose lifetime straddles any yield point, must be preserved while 
+the coroutine is suspended, and so they are hoisted into the coroutine environment.
+Note that this only moves their storage location, the lifetimes stay intact.  In the simplest case, 
+each hoisted variable gets its own unique storage space.
+A further optimization would be to overlay storage of variables which are not live simultaneously. 
+
+### No borrows across yield points
+Consider this code: 
+```rust
+    let a = vec![0; 100];
+    let b = a.iter();
+    yield c;
+    for x in b {
+        ...
+```
+Since both `a` and `b` are live across a yield point, they will be hoisted into the coroutine environment.
+Unfortunately, this means that the environment struct would store a reference to another part of self.
+This cannot be allowed because if the coroutine environment gets moved while it is suspended, 
+the internal pointer to `a` inside `b` would become invalid.  Thus, the compiler must emit an error.  
+The above does not prevent usage of references between yield points or having references to external objects. 
+
+### Cleanup
+When a regular closure goes out of scope, all of the closed-over variables implementing `Drop` are `drop()`'ped.  
+
+In coroutines, liveness of locals variables hoisted into the closure depends on the current state.
+If a coroutine runs to completion, all is well, because its hoisted locals will have been disposed of 
+in the course of normal execution.  However when a coroutine closure gets destroyed before reaching a return point, 
+some extra clean-up will be required:
+
+```rust
+impl Drop for CoroClosure1234 {
+    fn drop(&mut self) {
+        match (self.state) {
+            1 => { /* clean-up locals alive at yield point 1 */ },
+            2 => { /* clean-up locals alive at yield point 2 */ }
+            ...
+        }
+    }
+}
+
+```
+
+### Fn* traits
+Coroutines shall implement the `FnMut` trait:
```

Comment:
This information is quite crucial and I think it would be easier to read the RFC if somewhere near the beginning there was a sentence like “Coroutine is compiled to a closure which implements `FnMut(Args...) -> CoResult<Y, R>` trait. For example, the `coro1` coroutine is an `impl FnMut() -> CoResult<i32, ()>`”.